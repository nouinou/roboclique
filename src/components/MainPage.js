import React, { Component } from 'react';
import CardList from './card-list/CardList';
import Header from './shared/header/Header';
import Scroll from './shared/scroll/Scroll';
import ErrorBoundary from './shared/error-boundary/ErrorBoundary';

class MainPage extends Component {
	componentDidMount = () => this.props.fetchUsers();

	showMessage = (message) => (
		<div className="flex flex-col min-h-content justify-center items-center">
			<div className="robo-text text-2xl text-white max-w-lg">{message}</div>
		</div>
	);

	dataToShow = (isFetching, users) => {
		return isFetching ? (
			this.showMessage('Loading...')
		) : users && users.length ? (
			<Scroll>
				<ErrorBoundary>
					<CardList users={users} />
				</ErrorBoundary>
			</Scroll>
		) : (
			this.showMessage('No robot was found')
		);
	};

	filterRobots = () => {
		const { data, searchTerm } = this.props;
		return (
			data &&
			data.filter((user) =>
				user.fullname.toLowerCase().includes(searchTerm.toLowerCase())
			)
		);
	};

	render() {
		const { onSearchTermChanged, isFetching, error } = this.props;
		const robots = this.filterRobots();

		return (
			<div className="App text-center">
				<Header searchTermChanged={onSearchTermChanged} />
				{error
					? this.showMessage(
							'Something went wrong! Please try again later or contact the developer'
					  )
					: this.dataToShow(isFetching, robots)}
			</div>
		);
	}
}

export default MainPage;
