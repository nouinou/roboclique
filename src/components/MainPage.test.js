import React from 'react';
import MainPage from './MainPage';
import { shallow } from 'enzyme';

describe('MainPage', () => {
	let wrapper;
	const robots = [
		{
			fullname: 'Robo Cat',
		},
		{
			fullname: 'Miaw Bot',
		},
	];
	beforeEach(() => {
		const mockProps = {
			fetchUsers: jest.fn(),
			data: [],
			searchTerm: '',
			isFetching: false,
			error: null,
		};
		wrapper = shallow(<MainPage {...mockProps} />);
	});

	it('should render MainPage', () => {
		expect(wrapper).toMatchSnapshot();
	});

	describe('filterRobots', () => {
		it('return empty array', () => {
			const result = wrapper.instance().filterRobots();

			expect(result).toStrictEqual([]);
		});

		it('filters return array of length 1', () => {
			const mockProps = {
				fetchUsers: jest.fn(),
				data: robots,
				searchTerm: 'miaw',
				isFetching: false,
				error: null,
			};

			const expected = [
				{
					fullname: 'Miaw Bot',
				},
			];

			wrapper = shallow(<MainPage {...mockProps} />);
			const result = wrapper.instance().filterRobots();

			expect(result).toStrictEqual(expected);
		});

		it('filters return array of length 2', () => {
			const mockProps = {
				fetchUsers: jest.fn(),
				data: robots,
				searchTerm: 'bo',
				isFetching: false,
				error: null,
			};

			wrapper = shallow(<MainPage {...mockProps} />);
			const result = wrapper.instance().filterRobots();

			expect(result).toStrictEqual(robots);
		});

		it('filters return an empty array when no results was found', () => {
			const mockProps = {
				fetchUsers: jest.fn(),
				data: robots,
				searchTerm: 'meh',
				isFetching: false,
				error: null,
			};

			wrapper = shallow(<MainPage {...mockProps} />);
			const result = wrapper.instance().filterRobots();

			expect(result).toStrictEqual([]);
		});
	});

	describe('dataToShow', () => {
		it('should, if fetching, show loading message', () => {
			const loadingMessage = 'Loading...';
			const mockProps = {
				fetchUsers: jest.fn(),
				data: [],
				searchTerm: '',
				isFetching: true,
				error: null,
			};
			wrapper = shallow(<MainPage {...mockProps} />);
			const spy = jest.spyOn(wrapper.instance(), 'showMessage');

			wrapper.instance().showRobots();
			expect(spy).toHaveBeenCalledWith(loadingMessage);
		});

		it('should show No robots was found', () => {
			const notFound = 'No robot was found';
			const mockProps = {
				fetchUsers: jest.fn(),
				data: [],
				searchTerm: '',
				isFetching: false,
				error: null,
			};
			wrapper = shallow(<MainPage {...mockProps} />);
			const spy = jest.spyOn(wrapper.instance(), 'showMessage');

			wrapper.instance().showRobots();
			expect(spy).toHaveBeenCalledWith(notFound);
		});

		it('shoud show robots', () => {
			const mockProps = {
				fetchUsers: jest.fn(),
				data: robots,
				searchTerm: '',
				isFetching: false,
				error: null,
			};
			wrapper = shallow(<MainPage {...mockProps} />);
			const spy = jest.spyOn(wrapper.instance(), 'showScroll');

			wrapper.instance().showRobots();
			expect(spy).not.toHaveBeenCalled();
		});
	});

	describe('showMessage', () => {
		it('render loading message', () => {
			const loadingMessage = 'Loading...';
			const mockProps = {
				fetchUsers: jest.fn(),
				data: robots,
				searchTerm: '',
				isFetching: true,
				error: null,
			};

			wrapper = shallow(<MainPage {...mockProps} />);
			const result = wrapper.find('#message').text();

			expect(result).toBe(loadingMessage);
		});

		it('render no robot message', () => {
			const notFound = 'No robot was found';
			const mockProps = {
				fetchUsers: jest.fn(),
				data: [],
				searchTerm: '',
				isFetching: false,
				error: null,
			};

			wrapper = shallow(<MainPage {...mockProps} />);
			const result = wrapper.find('#message').text();

			expect(result).toBe(notFound);
		});

		it('render no robot message', () => {
			const errorMessage =
				'Something went wrong! Please try again later or contact the developer';
			const mockProps = {
				fetchUsers: jest.fn(),
				data: [],
				searchTerm: '',
				isFetching: false,
				error: {},
			};

			wrapper = shallow(<MainPage {...mockProps} />);
			const result = wrapper.find('#message').text();

			expect(result).toBe(errorMessage);
		});
	});
});
