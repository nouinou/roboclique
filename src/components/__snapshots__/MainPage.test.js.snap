// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should render MainPage 1`] = `
<div
  className="App text-center"
>
  <Header />
  <div
    className="flex flex-col min-h-content justify-center items-center"
  >
    <div
      className="robo-text text-2xl text-white max-w-lg"
      id="message"
    >
      No robot was found
    </div>
  </div>
</div>
`;
